compiler/env/Region.cpp

```diff
@@ -53,12 +53,11 @@ Region::~Region() throw()
     * Destroy all object instances that depend on the region
     * to manage their lifetimes.
     */
-   Destructable *lastDestructable = _lastDestructable;
-   while (lastDestructable)
+   Destroyer *lastDestroyer = _lastDestroyer;
+   while (lastDestroyer != NULL)
       {
-      Destructable * const currentDestructable = lastDestructable;
-      lastDestructable = currentDestructable->prev();
-      currentDestructable->~Destructable();
+      lastDestroyer->destroy();
+      lastDestroyer = lastDestroyer->prev();
```

Comment:
Just like the old code, I think it's unwise to assume the `prev` link is safe to access *after* the object has been destroyed: I think that behavior should be restored:
```
   Destroyer *lastDestroyer = _lastDestroyer;
   while (lastDestroyer != NULL)
      {
      Destroyer *prev = lastDestroyer->prev();
      lastDestroyer->destroy();
      lastDestroyer = prev;
      }
```

Reply:
`lastDestroyer` isn't destroyed here; it is a wrapper that contains a pointer to the object to be destroyed.

Reply:
Even so, this shouldn't assume that the implementation of `destroy()` didn't, say, clear the `prev` link.

Reply:
Well, the API contract for `destroy` in the current impl is to only destroy the containing object, not to reset the current `Destroyer`'s fields. The previous implementation destroyed the containing object (the `Destructible` object), whereas this implementation only destroys the contained class.

Reply:
> Even so, this shouldn't assume that the implementation of `destroy()` didn't, say, clear the `prev` link.

`Destroyer`, `TypedDestroyer`, and this loop are all just part of the implementation of `registerDestructor()`. There is no reason for them to evolve independently, and there is therefore no significant API surface between them, and we can know (not assume) precisely what `destroy()` does. This objection might as well be: this shouldn't assume that *the earlier part of the loop body* didn't clear the `prev` link—to which the answer would be that it plainly doesn't.